### PR TITLE
Image styles private files

### PR DESCRIPTION
**Pull request by joaogarin**
When trying to get private images we are running into an error of content being "leaked metadata found, check if rendered too early.." 

I guess something is triggering a render for the private images. Still trying to understand exactly where so I can provide better information.

I guess would also be nice to add a test for this as part of https://github.com/drupal-graphql/graphql/blob/8.x-4.x/tests/src/Kernel/DataProducer/Entity/Fields/Image/ImageDerivativeTest.php

### CODE REVIEW

File: src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php [hunk 1]
```diff
@@ -21,18 +25,69 @@
...
    * @return mixed
+   *   The file url.
    */
   public function resolve(FileInterface $entity, RefinableCacheableDependencyInterface $metadata) {
     $access = $entity->access('view', NULL, TRUE);
     $metadata->addCacheableDependency($access);
     if ($access->isAllowed()) {
-      return file_create_url($entity->getFileUri());
+      $context = new RenderContext();
+      $file_url = $this->renderer->executeInRenderContext($context, function () use ($entity, $image_style, $dimensions) {
```
Comment (klausi) [inline src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php]: you only use $entity, the other variables do not exist. I think that is why the tests are failing.

File: src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php [hunk 2]
```diff
@@ -21,18 +25,69 @@
...
+        return file_create_url($entity->getFileUri());
+      });
+      return $file_url;
```
Comment (klausi) [inline src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php]: this variable could be uninitialized here if access is denied. So we should set it first to NULL?

File: src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php (hunk 1 above)
Comment (joaogarin) [reply src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php]: Correct yes..my bad. fixed

File: src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php (hunk 2 above)
Comment (joaogarin) [reply src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php]: agree yes I think we should set it to null.if it fails I'd be ok with returning null

**Comment by joaogarin** [comment]
This could use some docs on like : 

"Working with image styles"

File: src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageDerivative.php [hunk 3]
```diff
@@ -26,15 +30,64 @@
...
   /**
    * @param \Drupal\file\FileInterface $entity
+   *   The file entity received.
+   * @param \Drupal\Core\Cache\RefinableCacheableDependencyInterface $metadata
+   *   The metadata object to add caching to objects.
    *
    * @param $style
    * @param \Drupal\Core\Cache\RefinableCacheableDependencyInterface $metadata
    *
    * @return mixed
```
Comment (klausi) [inline src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageDerivative.php:89]: should be `array|null`

File: src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageDerivative.php [hunk 4]
```diff
@@ -67,14 +120,17 @@ public function resolve(FileInterface $entity = NULL, $style, RefinableCacheable
       $image_style->transformDimensions($dimensions, $entity->getFileUri());
       $metadata->addCacheableDependency($image_style);
 
-      return [
-        'url' => $image_style->buildUrl($entity->getFileUri()),
-        'width' => $dimensions['width'],
-        'height' => $dimensions['height'],
-      ];
+      $context = new RenderContext();
```
Comment (klausi) [inline src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageDerivative.php:123]: Please add a code comment why we need the RenderContext here. Something like "// Prevent 'leaked metadata found, check if rendered too early' errors. @todo Find out why the render system is invoked at all?"

File: src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php [hunk 5]
```diff
@@ -21,21 +25,73 @@
...
+    $this->renderer = $renderer;
+  }
 
   /**
    * @param \Drupal\file\FileInterface $entity
+   *   The file entity.
    *
    * @param \Drupal\Core\Cache\RefinableCacheableDependencyInterface $metadata
    *
    * @return mixed
```
Comment (klausi) [inline src/Plugin/GraphQL/DataProducer/Entity/Fields/Image/ImageUrl.php:81]: should be `string|null`

**Review by klausi — COMMENTED** [review]
I guess this makes sense.

Can you set an xdebug breakpoint where this error message happens? Is that a way to track down when the early rendering already happened?

**Comment by benjy** [comment]
It renders early in PrivateStream::getExternalUrl. Would be great to have this PR merged, just wasted an hour on it.

```
  /**
   * {@inheritdoc}
   */
  public function getExternalUrl() {
    $path = str_replace('\\', '/', $this->getTarget());
    return Url::fromRoute('system.private_file_download', ['filepath' => $path], ['absolute' => TRUE, 'path_processing' => FALSE])->toString();
  }
```

**Comment by klausi** [comment]
Right, we have a more complete Pull request about this at #1039 , closing this in favor of that one.

**Comment by joaogarin** [comment]
awesome :tada:

**Pull request closed without merging**